feat(mi): create/view participant tags (M2-5861)

farmerpaul opened the pull request:
<!-- Use this template as a guide to describe your pull request, and adjust as necessary. -->
<!-- Include information that helps your peers review your updates and understand this    -->
<!-- repository's history of changes over time.                                           -->

<!-- Delete any options that are not relevant -->

- [x] Tests for the changes have been added

### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-5861](https://mindlogger.atlassian.net/browse/M2-5861)

<!-- Replace this with a high-level description of the features/functionality proposed in the pull request. -->

Adds create/view support for Participant tags to the admin panel. Includes:
- Tag dropdown in Add Participant popup
  - updated `SelectController` to add more universal support for existing `withChecked` prop
- Tags shown in the Participants table & Upgrade Account popup
  - created `ParticipantTagChip` component to be reused everywhere a tag is rendered
- Tags shown in the Take Now participant dropdown
- Tags shown in the Add Team Member popup (when role = "Reviewer")
  - refactored `TagsAutocompleteController` to be value-type-agnostic so it can render `ParticipantSnippet`s in the dropdown

### 📸 Screenshots

| Before                      | After                                 |
| -------------------------------------- | ------------------------------------- |
| <img width="434" alt="Screenshot 2024-05-07 at 4 48 24 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/de294278-a2c6-44c0-9da2-9c0a5a3fa59f"> | <img width="450" alt="Screenshot 2024-05-07 at 4 42 31 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/7ea677f1-3db7-4702-b5f0-b2ea35f396ca"> |
| <img width="1024" alt="Screenshot 2024-05-07 at 4 48 54 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/30d94675-3657-44ad-a18f-a9dd6f4a2a9e"> | <img width="1058" alt="Screenshot 2024-05-07 at 4 42 55 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/87700cd0-0901-4300-9ea0-32c76da7bf79"> |
| <img width="590" alt="Screenshot 2024-05-07 at 4 49 17 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/a7cc2316-58f1-42cd-972d-54358c94a22e"> | <img width="571" alt="Screenshot 2024-05-07 at 4 43 22 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/9436c757-36a0-4abb-85aa-47ccdb213169"> |
| <img width="614" alt="Screenshot 2024-05-07 at 4 49 35 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/dbf94046-1671-49af-9204-21aac9ee40da"> | <img width="588" alt="Screenshot 2024-05-07 at 4 43 47 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/ea94ba44-fffb-4979-a34d-82da58bdb0b6"> |
| <img width="769" alt="Screenshot 2024-05-07 at 4 50 25 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/c00c63a5-dfee-4011-913c-a3e38192b914"> | <img width="763" alt="Screenshot 2024-05-07 at 4 45 16 PM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/e3e3655b-8bf8-4a25-ac18-33503ac563c7"> |


### 🪤 Peer Testing

1. Navigate to Applet > Participants and click Add Participant.
2. Choose either Limited Account and click Next.
    **Expected outcome:** Tag selector should be available, with "No selection" selected by default.
3. Pick a tag and proceed to create the limited account.
    **Expected outcome:** New limited account should be listed with selected tag in Participants list.
4. Choose "Upgrade Account" from this account's action menu.
    **Expected outcome:** Participant's Tag should be rendered in the header of the Upgrade Account popup.
5. Close the Upgrade Account popup.
6. Navigate to Activities tab and choose Take Now for an activity.
7. Expand the "Who is this activity about?" dropdown.
    **Expected outcome:** Tag should be shown beside the new limited account in the dropdown list. **Note:** When you select this participant, the form field does not yet render the selected participant [as shown in Figma](https://www.figma.com/file/6BNx63peNVzrHAF1FoZ78x/Multi-Informant-2024?type=design&node-id=6026-299263&mode=design&t=J2LeCzoWTd4tlYik-4) as it was out of scope for this ticket.
8. Close the Take Now popup.
9. Navigate to Team tab and click Add Team Member.
10. Choose "Reviewer" for the Role field, and expand the Participants dropdown.
    **Expected outcome:** Tag should be shown beside the new limited account in the dropdown list.
11. Repeat the same steps above, but create a Full Account with a tag instead of a Limited Account.

farmerpaul committed:
M2-5861: add Tag selector to AddParticipantPopup

- update SelectController to support withChecked prop for
react-hook-form-controlled instances as well; hide icons in form field
in general, rather than applied specifically in the context of Legend
- add SelectController to AddParticipantForm to support tag selection
and save tag to the BE, and update tests
- add Flex styled component for end-justified content

farmerpaul committed:
M2-5861: show tags in Participants table

Also update tests, including fixing some incorrect testing mock data
and removing some warnings due to missing form defaults.

farmerpaul committed:
M2-5861: refactor rendered tag into component

Define ParticipantTagChip component to capture tag rendering logic from
Participants table, which will now be getting used in other places. Use
more strict typing (ParticipantTag) wherever tags are currently used.

farmerpaul committed:
M2-5861: add tags to Take Now participant selector

Adjust tag colour to translucent version so that it stands out against
selection background color in dropdown.

farmerpaul committed:
M2-5861: ♻️ tags selector to support rich options

Make TagsAutocompleteController value-type-generic (as MUI Autocomplete
was originally designed), so that custom options with more metadata can
have rich custom rendering. This way we can use ParticipantSnippet to
render tags in the Add Team Member form's Participants tag selector,
without any regressions to existing uses of TagsAutocompleteController.

Updated icons/spacings of TagsAutocompleteController,
LabeledUserDropdown and checkboxes to more precisely align with Figma.

JustinNusca commented on src/modules/Dashboard/components/ParticipantSnippet/ParticipantSnippet.styles.ts:
```diff
@@ -0,0 +1,7 @@
+
+export const StyledText = styled(StyledBodyLarge)`
+  ${ellipsisTextCss}
+`;
```
I'm not 100% certain how this compiles, but as `ellpsisTextCss` is already the output of a `css` function, it may be better to write this as:

```suggestion
export const StyledText = styled(StyledBodyLarge)(ellipsisTextCss);
```

… to reuse the existing set of styles, rather than creating a new template literal to be parsed, consisting solely of the embedded styles. This goes for the same usage [here](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1706/files#diff-a55c2a9f8054d4d50378b3b63c84b25534a823578581f165d43c7616572fce8cR152-R155).

farmerpaul replied:
Sweet, thanks! I didn't realize this type of string literal syntax was just passing a string as an argument to a callback. 🤦🏻‍♂️

JustinNusca commented on src/modules/Dashboard/components/ParticipantTagChip/ParticipantTagChip.types.ts:
```diff
@@ -0,0 +1,6 @@
+import { ParticipantTag } from 'shared/consts';
+
+export type ParticipantTagProps = {
```
Nit: Let's name this consistently with the component these props are for, either:

`ParticipantTagChipProps` or rename the component to `ParticipantTag`.

farmerpaul replied:
Totally, thanks for noticing… It was a relic from when the component was still called `ParticipantTag`, but that identifier conflicted with the enum `ParticipantTag` and I wanted to avoid confusion.

JustinNusca commented on src/shared/components/FormComponents/TagsAutocompleteController/TagsAutocompleteController.tsx:
```diff
@@ -143,6 +159,9 @@ export const TagsAutocompleteController = <T extends FieldValues>({
               },
+              paper: {
+                sx: { maxHeight: '25.6rem' }, // Select All row + 4 rows
+              },
```
Given that we're already passing a `PaperComponent` prop, with a component that renders a pre-configured MUI `Paper` component (ie, with custom children), do you think it would make sense to just move this so all of our paper customizations are co-located together? EG, update [this line](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1706/files#diff-cbe500fb4bfc7e7c4434161161744cdb7a3d8b8f876ec6917882047ecb6852f6R127) to:

```jsx
<Paper {...restPaperProps} sx={{ maxHeight: '25.6rem' }}>
```

… rather than scattering the config for this component out to this line?

farmerpaul replied:
Totally, great eye again! 🦅

JustinNusca reviewed (approved):
A few tiny suggestions, but I'm pre-approving regardless — it's looking good and working well from my testing. 👍

farmerpaul committed:
M2-5861: update tag naming to align with API

Tags are going to be in Title case in the BE; also the "owner" tag has
been removed.

farmerpaul committed:
M2-5861: PR change requests

sultanofcardio commented on src/modules/Dashboard/components/TakeNowModal/LabeledDropdown/LabeledUserDropdown.tsx:
```diff
@@ -19,6 +25,7 @@ export const LabeledUserDropdown = ({
   onChange,
   handleSearch,
   debounce,
+  disabled,
```
Why the explicit disabled prop here?

farmerpaul replied:
It's used to affect the colour of the `popupIcon` SVG. 🦅
